File: bindings/java/src/main/java/com/moofile/Document.java
package com.moofile;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * A MooFile document — an insertion-ordered map of field names to values.
 *
 * <p>Values are the plain Java types {@link Json} produces: {@link String},
 * {@link Long}, {@link Double}, {@link Boolean}, {@link List}, nested
 * {@link Map}, and null.
 *
 * <pre>{@code
 * Document doc = new Document()
 *     .put("name", "Alice")
 *     .put("age", 30)
 *     .put("tags", List.of("admin", "ops"));
 * }</pre>
 */
public class Document {

    private final Map<String, Object> data;

    public Document() {
        this.data = new LinkedHashMap<>();
    }

    public Document(Map<String, Object> data) {
        this.data = data == null ? new LinkedHashMap<>() : new LinkedHashMap<>(data);
    }

    /** Build a document from alternating key/value arguments. */
    public static Document of(Object... keyValuePairs) {
        if (keyValuePairs.length % 2 != 0) {
            throw new IllegalArgumentException("expected alternating key/value arguments");
        }
        Document d = new Document();
        for (int i = 0; i < keyValuePairs.length; i += 2) {
            d.put(String.valueOf(keyValuePairs[i]), keyValuePairs[i + 1]);
        }
        return d;
    }

    /** Set a field. Returns this, for chaining. */
    public Document put(String key, Object value) {
        data.put(key, value);
        return this;
    }

    public Object get(String key) {
        return data.get(key);
    }

    /** Get a field as a string, or null if absent. */
    public String getString(String key) {
        Object v = data.get(key);
        return v == null ? null : v.toString();
    }

    /**
     * Get a numeric field as a long.
     *
     * @throws MooFileException if the field is absent or not a number
     */
    public long getLong(String key) {
        Object v = data.get(key);
        if (v instanceof Number n) return n.longValue();
        throw new MooFileException("field '" + key + "' is not a number: " + v);
    }

    /**
     * Get a numeric field as a double.
     *
     * @throws MooFileException if the field is absent or not a number
     */
    public double getDouble(String key) {
        Object v = data.get(key);
        if (v instanceof Number n) return n.doubleValue();
        throw new MooFileException("field '" + key + "' is not a number: " + v);
    }

    /** Get a boolean field, defaulting to false when absent. */
    public boolean getBoolean(String key) {
        return Boolean.TRUE.equals(data.get(key));
    }

    /** Get a list field, or null if absent. */
    @SuppressWarnings("unchecked")
    public List<Object> getList(String key) {
        Object v = data.get(key);
        return v instanceof List ? (List<Object>) v : null;
    }

    /** Get a nested object field as a Document, or null if absent. */
    @SuppressWarnings("unchecked")
    public Document getDocument(String key) {
        Object v = data.get(key);
        return v instanceof Map ? new Document((Map<String, Object>) v) : null;
    }

    /** The document's {@code _id}, or null before insertion. */
    public String id() {
        return getString("_id");
    }

    public boolean containsKey(String key) { return data.containsKey(key); }
    public Object remove(String key)       { return data.remove(key); }
    public int size()                      { return data.size(); }
    public boolean isEmpty()               { return data.isEmpty(); }
    public java.util.Set<String> keySet()  { return data.keySet(); }

    /** The backing map. Mutating it mutates the document. */
    public Map<String, Object> asMap() { return data; }

    public String toJson() { return Json.write(data); }

    /** Parse a JSON object into a document. */
    public static Document parse(String json) {
        return new Document(Json.parseObject(json));
    }

    @Override
    public String toString() { return toJson(); }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        return o instanceof Document other && data.equals(other.data);
    }

    @Override
    public int hashCode() { return data.hashCode(); }
}

File: bindings/csharp/Moofile/Document.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Moofile;

/// <summary>Raised for every MooFile failure.</summary>
public class MooFileException : Exception
{
    public MooFileException(string message) : base(message) { }
    public MooFileException(string message, Exception inner) : base(message, inner) { }
}

/// <summary>
/// A MooFile document — an insertion-ordered map of field names to values.
/// </summary>
/// <remarks>
/// Values read back from the database are plain CLR types (string, long,
/// double, bool, null, <see cref="List{T}"/>, nested <see cref="Document"/>)
/// rather than <c>JsonElement</c>, so <c>doc["age"]</c> compares and casts the
/// way you would expect.
/// </remarks>
public class Document : IEnumerable<KeyValuePair<string, object?>>
{
    private readonly Dictionary<string, object?> _data = new();
    private readonly List<string> _order = new();

    public Document() { }

    public Document(IDictionary<string, object?> source)
    {
        foreach (var kvp in source) this[kvp.Key] = kvp.Value;
    }

    /// <summary>Build a document from alternating key/value arguments.</summary>
    public static Document Of(params object?[] keyValuePairs)
    {
        if (keyValuePairs.Length % 2 != 0)
            throw new ArgumentException("expected alternating key/value arguments");

        var d = new Document();
        for (int i = 0; i < keyValuePairs.Length; i += 2)
        {
            d[Convert.ToString(keyValuePairs[i]) ?? ""] = keyValuePairs[i + 1];
        }
        return d;
    }

    public object? this[string key]
    {
        get => _data.TryGetValue(key, out var v) ? v : null;
        set
        {
            if (!_data.ContainsKey(key)) _order.Add(key);
            _data[key] = value;
        }
    }

    /// <summary>Set a field and return this, for chaining.</summary>
    public Document Set(string key, object? value)
    {
        this[key] = value;
        return this;
    }

    public bool ContainsKey(string key) => _data.ContainsKey(key);

    public bool Remove(string key)
    {
        _order.Remove(key);
        return _data.Remove(key);
    }

    public int Count => _data.Count;
    public IReadOnlyList<string> Keys => _order;

    /// <summary>The document's <c>_id</c>, or null before insertion.</summary>
    public string? Id => GetString("_id");

    public string? GetString(string key) => this[key]?.ToString();

    public long GetLong(string key) => this[key] switch
    {
        long l => l,
        int i => i,
        double d => (long)d,
        _ => throw new MooFileException($"field '{key}' is not a number: {this[key]}"),
    };

    public double GetDouble(string key) => this[key] switch
    {
        double d => d,
        long l => l,
        int i => i,
        _ => throw new MooFileException($"field '{key}' is not a number: {this[key]}"),
    };

    public bool GetBoolean(string key) => this[key] is true;

    public List<object?>? GetList(string key) => this[key] as List<object?>;

    public Document? GetDocument(string key) => this[key] as Document;

    public IEnumerator<KeyValuePair<string, object?>> GetEnumerator()
    {
        foreach (var key in _order) yield return new(key, _data[key]);
    }

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

    // -----------------------------------------------------------------
    // JSON
    // -----------------------------------------------------------------

    public string ToJson()
    {
        var node = ToNode(this);
        return node?.ToJsonString() ?? "{}";
    }

    /// <summary>Convert any supported value into a JsonNode tree.</summary>
    internal static JsonNode? ToNode(object? value)
    {
        switch (value)
        {
            case null:
                return null;
            case Document doc:
            {
                var obj = new JsonObject();
                foreach (var (key, v) in doc) obj[key] = ToNode(v);
                return obj;
            }
            case IDictionary<string, object?> map:
            {
                var obj = new JsonObject();
                foreach (var kvp in map) obj[kvp.Key] = ToNode(kvp.Value);
                return obj;
            }
            case string s:
                return JsonValue.Create(s);
            case bool b:
                return JsonValue.Create(b);
            case int or long or short or byte:
                return JsonValue.Create(Convert.ToInt64(value));
            case float or double or decimal:
                return JsonValue.Create(Convert.ToDouble(value));
            case JsonNode node:
                return node.DeepClone();
            case IEnumerable seq:
            {
                var arr = new JsonArray();
                foreach (var item in seq) arr.Add(ToNode(item));
                return arr;
            }
            default:
                return JsonValue.Create(value.ToString());
        }
    }

    /// <summary>Parse a JSON object into a document.</summary>
    public static Document Parse(string json)
    {
        JsonNode? node;
        try
        {
            node = JsonNode.Parse(json);
        }
        catch (JsonException e)
        {
            throw new MooFileException($"invalid JSON: {e.Message}", e);
        }

        if (node is not JsonObject obj)
            throw new MooFileException("expected a JSON object");

        return (Document)FromNode(obj)!;
    }

    /// <summary>
    /// Convert a JsonNode tree into plain CLR values. Numbers become long
    /// when they are integral and double otherwise, so scores stay doubles
    /// while counts stay integers.
    /// </summary>
    internal static object? FromNode(JsonNode? node)
    {
        switch (node)
        {
            case null:
                return null;
            case JsonObject obj:
            {
                var doc = new Document();
                foreach (var kvp in obj) doc[kvp.Key] = FromNode(kvp.Value);
                return doc;
            }
            case JsonArray arr:
            {
                var list = new List<object?>(arr.Count);
                foreach (var item in arr) list.Add(FromNode(item));
                return list;
            }
            case JsonValue val:
            {
                if (val.TryGetValue<bool>(out var b)) return b;
                if (val.TryGetValue<string>(out var s)) return s;
                if (val.TryGetValue<long>(out var l)) return l;
                if (val.TryGetValue<double>(out var d)) return d;
                return val.ToJsonString();
            }
            default:
                return null;
        }
    }

    public override string ToString() => ToJson();
}

/// <summary>A document paired with its similarity or relevance score.</summary>
public class SearchResult
{
    public Document Doc { get; init; } = new();
    public double Score { get; init; }

    public override string ToString() => $"{Score}: {Doc}";
}
